### PR TITLE
Expose module forces

### PR DESCRIPTION
**Pull request by BBScholar**
Addresses: https://github.com/SleipnirGroup/Choreo/issues/479

Related to: https://github.com/SleipnirGroup/Choreo/pull/480

I feel like this solution isn't necessarily robust. I think I will need to add support variable number of modules to remain consistent.

### CODE REVIEW

File: include/trajopt/trajectory/HolonomicTrajectorySample.h [hunk 1]
```diff
@@ -32,6 +34,12 @@ class TRAJOPT_DLLEXPORT HolonomicTrajectorySample {
   /// The angular velocity.
   double angularVelocity = 0.0;
 
+  /// The force on each module in the X direction.
+  std::vector<double> moduleForcesX{};
+
+  /// The force on each module in the Y direction.
+  std::vector<double> moduleForcesY{};
```
Comment (calcmogul) [inline include/trajopt/trajectory/HolonomicTrajectorySample.h:41]:
```suggestion
  std::vector<double> moduleForcesX;

  /// The force on each module in the Y direction.
  std::vector<double> moduleForcesY;
```

File: src/trajoptlibrust.cpp [hunk 2]
```diff
@@ -261,6 +262,13 @@ void SwervePathBuilderImpl::sgmt_polygon_obstacle(size_t from_idx,
 
 HolonomicTrajectorySample _convert_holonomic_trajectory_sample(
     const trajopt::HolonomicTrajectorySample& sample) {
+
+  rust::Vec<double> fx, fy;
+
+  // copy data into rust vecs
+  std::copy(sample.moduleForcesX.begin(), sample.moduleForcesX.end(), std::back_inserter(fx));
+  std::copy(sample.moduleForcesY.begin(), sample.moduleForcesY.end(), std::back_inserter(fy));
```
Comment (calcmogul) [inline src/trajoptlibrust.cpp:270]:
```suggestion
  rust::Vec<double> fx;
  std::copy(sample.moduleForcesX.begin(), sample.moduleForcesX.end(), std::back_inserter(fx));
  
  rust::Vec<double> fy;
  std::copy(sample.moduleForcesY.begin(), sample.moduleForcesY.end(), std::back_inserter(fy));
```

File: src/trajoptlibrust.cpp [hunk 3]
```diff
@@ -269,6 +277,8 @@ HolonomicTrajectorySample _convert_holonomic_trajectory_sample(
       .velocity_x = sample.velocityX,
       .velocity_y = sample.velocityY,
       .angular_velocity = sample.angularVelocity,
+      .module_forces_x = fx,
+      .module_forces_y = fy,
```
Comment (calcmogul) [inline src/trajoptlibrust.cpp:281]: Can these be std::moved into place?

File: include/trajopt/trajectory/HolonomicTrajectory.h [hunk 4]
```diff
@@ -2,11 +2,13 @@
 
 #pragma once
 
+#include <algorithm>
```
Comment (calcmogul) [inline include/trajopt/trajectory/HolonomicTrajectory.h:5]: This doesn't appear to be used in HolonomicTrajectory.h.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: include/trajopt/trajectory/HolonomicTrajectorySample.h (hunk 1 above)
Comment (BBScholar) [reply include/trajopt/trajectory/HolonomicTrajectorySample.h:41]: This change will prevent the default constructor from being `constexpr` it seems.

File: src/trajoptlibrust.cpp (hunk 3 above)
Comment (BBScholar) [reply src/trajoptlibrust.cpp]: Seems so.

File: include/trajopt/trajectory/HolonomicTrajectorySample.h (hunk 1 above)
Comment (BBScholar) [reply include/trajopt/trajectory/HolonomicTrajectorySample.h:41]: image

**Comment by BBScholar** [comment]
Not really sure how to fix the windows builds. I don't have a windows machine to test on ATM.

**Comment by BBScholar** [comment]
MSVC doesn't support `constexpr` vector operations it seems.

**Comment by calcmogul** [comment]
VS 2019 added support for constexpr std::vector: https://en.cppreference.com/w/cpp/compiler_support/20

Make sure you're using a new enough version of VS 2022.

**Comment by BBScholar** [comment]
I'm specifically talking about the Windows - Casadi build action here on github. All the other windows builds are passing.

**Comment by calcmogul** [comment]
Looks like this is caused by the Windows CasADi backend using an ancient MinGW version.

**Comment by calcmogul** [comment]
I'm debating removing the CasADi backend. We can't even use it in Choreo right now because MinGW's linker is too old to handle the version of Rust we use.

**Pull request merged**